### PR TITLE
Firefly 37 recognizing s_region in a table colum

### PR DESCRIPTION
**Pull request by cwang2016**
https://jira.ipac.caltech.edu/browse/FIREFLY-42 (this PR is for FIREFLY-42, not 37)
https://jira.lsstcorp.org/browse/DM-18510
this PR includes development on

* create two drawing layers for the table has columns compliant to ObsCore table with s_region column:
    * observation center layer contain point objects for s_ra & s_dec column, like normal catalog layer
    * observation region layer contain region objects for s_region column
* the layer dialog shows column info for both observation center and region layers.
* the layer dialog shows buttons on 'all' regions, 'selected' regions and 'highlighted' region for the 
observation region layer
* the layer dialog shows an error icon if any of the region for ‘all’, ’selected’ or ‘highlighted’ selection has invalid or unsupported values in the description.  
* the point or the region object on either layer can be clicked to highlighted, and the the table highlight will appear accordingly. 
* the point on observation center layer can be either selected or filtered using ‘select area’ tool. 
* the region on observation region layer can not be filtered or selected by using ‘select area’ tool. 
* the selection or filtering made on observation center layer will reflect to the region layer and 
the corresponding rows of the table. 
* the selection or filters made on the table will be visible on both observation center and region layers. 

test 
k8s build 
https://irsawebdev9.ipac.caltech.edu/firefly-37/firefly/firefly-dev.html
* do tap search & cone search (m32, 5 arcsec) on the following cases, 
CADC => ivoa => ivoa.Obscore, (all rows with valid s_region value) or  
MAST => ivoa => ivoa.Obscore (some rows with unsupported s_region value)
the coverage view will contain two layers (one for region and one for center) for the table with s_region column. 
* try filter & select on either layer by using 'select area' tool 
* try highlight  by clicking the point or the border of the circle or polygon
* open layer dialog, try 
    * turn on/off the center or the region layer, 
    * for region layer, try the button on different region selection, and check the error message if there is.

### CODE REVIEW

File: src/firefly/js/visualize/draw/CanvasWrapper.jsx [hunk 1]
```diff
@@ -55,7 +55,7 @@ function makeCanvasLayers(drawLayer,drawer,w,h, idx, plotId) {
                         id={mainId}
                         ref={(c) => drawer.setPrimCanvas(c,w,h)}/>);
 
-    if (canSelect) {
+    if (canSelect || canShowSelect) {
```
Comment (robyww) [inline src/firefly/js/visualize/draw/CanvasWrapper.jsx]: glad you caught this.

**Comment by robyww** [comment]
Tested the UI and it look very nice!

**Comment by robyww** [comment]
Before you merge I do want to discuss with you how we are recognizing a table with s_region.  I think you might be being too strict.

File: src/firefly/js/visualize/ui/TriViewImageSection.jsx [hunk 2]
```diff
@@ -191,7 +192,8 @@ function handleNewTable(layoutInfo, action) {
     let {coverageLockedOn, showFits, showMeta, showCoverage, selectedTab, metaDataTableId} = images;
     const isMeta = isMetaDataTable(tbl_id);
     
-    if ((isMeta || isCatalog(tbl_id)) && showTables ) {
+
+    if ((isMeta || isCatalog(tbl_id) || isTableWithRegion(tbl_id)) && showTables ) {
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/ui/TriViewImageSection.jsx]: yes. the merge is done.

File: src/firefly/js/visualize/ui/TriViewImageSection.jsx [hunk 3]
```diff
@@ -230,12 +232,14 @@ function onActiveTable (layoutInfo, action) {
     // check for catalog or meta images
     const anyHasCatalog= hasCatalogTable(tblList);
     const anyHasMeta= hasMetaTable(tblList);
+    const anyHasRegion = hasRegionTable(tblList);
+
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/ui/TriViewImageSection.jsx]: done

File: src/firefly/js/visualize/ui/TriViewImageSection.jsx [hunk 4]
```diff
@@ -230,12 +232,14 @@ function onActiveTable (layoutInfo, action) {
...
 
     // if (coverageLockedOn) {
     //     coverageLockedOn= anyHasCatalog || anyHasMeta;
     // }
 
-    if (anyHasCatalog || anyHasMeta) {
+    if (anyHasCatalog || anyHasMeta || anyHasRegion) {
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/ui/TriViewImageSection.jsx]: done

[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/ui/TriViewImageSection.jsx]: done

File: src/firefly/js/visualize/draw/ShapeDataObj.js [hunk 5]
```diff
@@ -258,6 +258,9 @@ function makeDrawParams(drawObj,def={}) {
     };
 }
 
+const getCenter = (obj) => {
+    draw.getCenterPt(obj);
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/draw/ShapeDataObj.js]: removed

File: src/firefly/js/visualize/draw/ShapeDataObj.js [hunk 6]
```diff
@@ -295,12 +298,14 @@ const draw=  {
...
+
         if (drawObj.sType === ShapeType.Line ) return distToLine(drawObj.pts, plot, pt);
         if (drawObj.sType === ShapeType.Polygon) return distanceToPolygon(drawObj, plot, pt);
+        if (drawObj.sType === ShapeType.Circle) return distanceToCircle(drawObj, plot, pt);
+        if (drawObj.sType === ShapeType.Rectangle) return distanceToRectangle(drawObj, plot, pt);
 
-        const testPt = plot.getScreenCoords(this.getCenterPt(drawObj));
-
+        if (isScreenPtInRegion(drawObj, pt, plot).inside) return 0;
+        const testPt = plot.getScreenCoords(getCenter(drawObj));
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/draw/ShapeDataObj.js]: draw.getCenterPt works

File: src/firefly/js/visualize/draw/ShapeDataObj.js [hunk 7]
```diff
@@ -1795,27 +1800,93 @@ export function distToLine(pts, cc, pt) {
...
+        const nIdx = (idx+1)%len;
+        const d = distToLine([corners[idx], corners[nIdx]], cc, spt);
+
+        if (d < prev) {
+            prev = d;
+        }
+        return prev;
+    }, dist);
+}
+
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/visualize/draw/ShapeDataObj.js]: move generic distanceToCircle, distanceToLine, distanceToPolygon to VisUtil. 
Keep distanceToRectangle which is drawObj based.

File: src/firefly/js/ui/PopupUtil.jsx [hunk 8]
```diff
@@ -70,24 +70,27 @@ export function showOptionsPopup({content, title='Options', modal = false, show=
...
  * @param {string} title
+ * @param {boolean} isCopyable
  * @return {object}
  */
-export function showInfoPopup(content, title='Information') {
+export function showInfoPopup(content, title='Information', isCopyable = false) {
     var results= (
         <PopupPanel title={title} >
-            {makeContent(content)}
+            {makeContent(content, isCopyable)}
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/ui/PopupUtil.jsx]: yes

**Pull request merged**